proposals/simd-0182-conditional-cu-metering.md

```diff
@@ -0,0 +1,114 @@
+track the exact number of executed sBPF instructions for consensus is costly and
+unnecessary. Such precision is not essential for protocol-level consensus,
+especially since these cases are rare.
+
+Instead of mandating implementation-specific work to handle exceptions, we
+propose the following clarification in the protocol:
+
+- For successful execution of a Basic Block (i.e., the block exits at the last
+ sBPF instruction), the deterministic CU cost of the block will be charged to
+the transaction’s CU meter. This ensures that CU consumption for successful
+transactions is accurately accounted for.
+- In the event of irregular failure, where execution aborts from the middle of
+basic block, the requested CUs for the transaction will be charged to the CU
+meter. This allows for a simple and efficient fallback mechanism that avoids the
+need for tracking the exact number of executed instructions up to the point of
+failure.
+
+By adopting this approach, the protocol avoids the overhead of requiring precise
+instruction-level CU tracking for transactions that fail. Instead, the requested
+CU limit of the transaction will be used, simplifying the handling of
+irregularly failed transactions while still maintaining consensus.
+
+## Alternatives Considered
+
+None
+
+## New Terminology
+
+- [Basic Block](https://en.wikipedia.org/wiki/Basic_block): In the context of
+  JIT execution and BPF processing, a Basic Block is a sequence of BPF
+instructions that forms a single, linear flow of control with no loops or
+conditional branches except for the entry and exit points. It represents a
+segment of code where execution starts at the first instruction and proceeds
+sequentially through to the last instruction without deviation. The Basic Block
+is characterized by its predictable execution path, allowing for efficient
+budget checks and optimizations, as its Compute Unit (CU) cost can be determined
+before execution and verified at the end of the block.
+
+- Irregular transaction failure: A rare case that a Transaction execution aborts
+in the middle of executing basic block, results in consuming all requested CUs.
+
+## Detailed Design
+
+If VM execution returns any error except `SyscallError`, transaction's CU meter
+should be fully depleted, in another words, all requested CUs are consumed;
+otherwise consumes the actual executed CUs.
+
+## Impact
+
+None
+
+## Security Considerations
+
+One potential issue with using requested CUs in the case of failed transactions
+is the risk of transactions with grossly large CU requests consuming an
+excessive portion of the block's CU limit. This could effectively cause a
+denial-of-service effect by preventing legitimate transactions from being
+included in the block. To mitigate this risk, it is recommended that this
+proposal be implemented after SIMD-172 is deployed, which removes the
+possibility of accidentally requesting an excessively large number of CUs.
```

Comment:
I don't think this is a real security concern. Transactions always pay for the amount of compute units they are requesting. If they get scheduled and fail and consume a large amount of the block cost then other transaction senders should increase their priority fees to get scheduled first.